### PR TITLE
Add suggest Vec::push when the wrong method is used

### LINKED ISSUE #87212: Add hint to correct `Vec::append` into `Vec::push` when the wrong method is used.
Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a6f4c87ca812c201bfb57ef861c4d919

```rust
struct Thing;

fn main() {
    let mut stuff: Vec<Thing> = Vec::new();

    stuff.append(Thing);
}
```

The current output is:

```
error[E0308]: mismatched types
 --> src/main.rs:6:18
  |
6 |     stuff.append(Thing);
  |                  ^^^^^ expected `&mut Vec<Thing>`, found struct `Thing`
  |
  = note: expected mutable reference `&mut Vec<Thing>`
                        found struct `Thing`
```

Ideally the output should look like:

```
error[E0308]: mismatched types
 --> src/main.rs:6:18
  |
6 |     stuff.append(Thing);
  |                  ^^^^^
  |                  |
  |                  expected `&mut Vec<Thing>`, found struct `Thing`
  |                  help: use `push` to add a single element
  |
  = note: expected mutable reference `&mut Vec<Thing>`
                        found struct `Thing`
```

There should be some kind of suggestion to use `push` when a value `T` is incorrectly `append`-ed onto a `Vec<T>`. This likely only trips up newcomers to rust, but I know a fair number of python users who had this as their first error message. There are likely some some other collections this could apply to, but `Vec` would definitely constitute most of these mistakes.

If nobody is immediately opposed to this change, I would like to try to implement this. I would need some help with _where_ to make this change, but this seems like a fair introduction into editing compiler output.

It may also be helpful to just add a section mentioning `push` to the `append` function docs.

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #87212

`append` is commonly used in Python, suggest `push` if possible.

### CODE REVIEW

File: compiler/rustc_infer/src/infer/error_reporting/suggest.rs [hunk 1]
```diff
@@ -289,6 +289,61 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
...
+        let mut finder = MethodCallVisitor { method: None, err_span: span, recv: None };
+        finder.visit_body(body);
+
+        if let Some(typecheck) = self.typeck_results.as_ref()
+            && let Some(method_call) = finder.method
+            && let Some(vec_def_id) = self.tcx.get_diagnostic_item(sym::Vec)
+            && let Some(recv_ty) = finder.recv.and_then(|recv| typecheck.expr_ty_opt(recv))
+            && let ty::Adt(def, _) = recv_ty.kind()
+            && def.did() == vec_def_id
+            && (recv_ty.contains(*found) || recv_ty.has_infer_types())
```
Comment (chenyukang) [inline compiler/rustc_infer/src/infer/error_reporting/suggest.rs:336]: maybe it's better to check coerce between `recv_ty` and `expected ty`?

**Comment by oli-obk** [comment]
While this suggestion is really precise and targeted, would it be sufficient to use `#[rustc_confusables(...)]` on `push`?

cc https://github.com/rust-lang/rust/issues/59450

**Comment by jieyouxu** [comment]
> While this suggestion is really precise and targeted, would it be sufficient to use `#[rustc_confusables(...)]` on `push`?
> 
> cc #59450

I think estebank is also adding/improving some infrastructure for this, cf. https://github.com/rust-lang/rust/pull/120730?

**Comment by chenyukang** [comment]
emm, the issue was not linked with a PR 😂
closing it.

**Comment by estebank** [comment]
@chenyukang yeah, it's a shame we have some overlapping tickets that are not linked to each other leading to situations like this one :-/

**Pull request closed without merging**